File: Cracking-the-Coding-Interview/1-Implementations/Data-Structures/Linked-List/Singly-Linked-List/Python/singly.py
import warnings


class LinkedList:

    class Node:
        def __init__(self, value=0):
            self.value = value
            self.next = None


    def __init__(self, value=None):
        self.head = None if not value else LinkedList.Node(value)
        self._size = 0 if not self.head else 1


    def insert_head(self, value=None):
        """
        Inserts a new node as the head of the linked list.

        O(1) time complexity.

        Args:
            value: The value to insert.

        Returns:
            True if the insertion was successful.
        """

        if value == None:
            warnings.warn('Missing value during head insertion')
            return False
        
        try:
            new_head = LinkedList.Node(value)
            new_head.next = self.head
            self.head = new_head
            self._size += 1
            return True
        except Exception as error:
            print(f"Error occurred while inserting a new head node: {str(error)}")
            raise
    
    
    def insert_tail(self, value=None):
        """
        Inserts a new node as the tail of the linked list.

        O(n) time complexity.

        Args:
            value: The value to insert.

        Returns:
            True if the insertion was successful.
        """

        if value == None:
            warnings.warn('Missing value during tail insertion')
            return False
        
        try:

            if self._size == 0:
                return self.insert_head(value)
            
            curr = self.head

            while curr is not None and curr.next is not None:
                curr = curr.next
            
            new_tail = LinkedList.Node(value)
            curr.next = new_tail
            self._size += 1
            return True
        except Exception as error:
            print(f"Error occurred while inserting a new head node: {str(error)}")
            raise
    
    
    def insert_at_index(self, value=None, index=None):
        """
        Inserts a new node into the given index.

        O(n) time complexity.

        Args:
            value: The value to insert.
            index: Index to insert into

        Returns:
            True if the insertion was successful.
        """

        if value == None:
            warnings.warn('Missing value during index insertion')
            return False
        
        if index == None:
            warnings.warn('Missing index during index insertion')
            return False
        
        if index < 0 or index >= self._size:
            warnings.warn('Invalid index during index insertion')
            return False
        
        try:
            if index == 0:
                return self.insert_head(value)
            
            if index == self._size - 1:
                return self.insert_tail(value)
            
            curr = self.head

            for i in range(index - 1):
                curr = curr.next
            
            new_idx_node = LinkedList.Node(value)
            prev_idx_node = curr.next

            curr.next = new_idx_node
            new_idx_node.next = prev_idx_node
            self._size += 1
            return True
        except Exception as error:
            print(f"Error occurred while inserting a new node at index {index}: {str(error)}")
            raise
    
    
    def delete_head(self):
        """
        Deletes the head node.

        O(1) time complexity.

        Returns:
            True if the deletion was successful.
        """

        if not self.head:
            return False
        
        try:
            new_head = self.head.next
            self.head.next = None
            self.head = new_head
            self._size -= 1
            return True
        except Exception as error:
            print(f"Error occurred while deleting head node: {str(error)}")
            raise
    
    
    def delete_tail(self):
        """
        Deletes the tail node.

        O(n) time complexity.

        Returns:
            True if the deletion was successful.
        """

        if not self.head:
            return False
        
        if self._size == 1:
            self.head = None
            self._size -= 1
            return True
        
        try:
            curr = self.head

            for i in range(self._size - 2):
                curr = curr.next

            curr.next = None
            self._size -= 1
            return True
        except Exception as error:
            print(f"Error occurred while deleting head node: {str(error)}")
            raise
    
    
    def delete_at_index(self, index=None):
        """
        Deletes a node at the given index.

        O(n) time complexity.

        Args:
            index: Node index to delete

        Returns:
            True if the deletion was successful.
        """

        if index == None:
            warnings.warn('Missing index during index deletion')
            return False
        
        if index < 0 or index >= self._size:
            warnings.warn('Invalid index during index deletion')
            return False
        
        try:
            if index == 0:
                return self.delete_head()
            
            if index == self._size - 1:
                return self.delete_tail()
            
            curr = self.head

            for i in range(index - 1):
                curr = curr.next
            
            deleted_node = curr.next
            next_node = deleted_node.next

            deleted_node.next = None
            curr.next = next_node
            self._size -= 1
            return True
        except Exception as error:
            print(f"Error occurred while deleting a node at index {index}: {str(error)}")
            raise
    
    
    def has_value(self, value=None):
        """
        Searches the list for the given value.

        O(n) time complexity.

        Args:
            value: node value to search for

        Returns:
            True if the list contains the value.
        """

        if value == None:
            warnings.warn('Invalid search value')
            return False
        
        if self._size == 0:
            return False
        
        try:
            curr = self.head

            while curr is not None:
                if curr.value == value:
                    return True
                curr = curr.next

            return False
        except Exception as error:
            print(f"Error occurred while searching for value {value}: {str(error)}")
            raise


    def get_size(self):
        """
        Returns the size of the linked list

        O(1) time complexity.

        Returns:
            Number of nodes in the list (number)
        """
        return self._size


    def print_list(self):
        """
        Prints the linked list to the console

        O(n) time complexity.
        """
        if not self.head:
            print('Empty List')
            return
        
        curr = self.head
        output = '\nLinked List: '

        while curr is not None:
            output += f"{curr.value} -> "
            curr = curr.next

        output += 'NULL\n'
        print(output)

File: Cracking-the-Coding-Interview/1-Implementations/Data-Structures/Linked-List/Singly-Linked-List/JavaScript/singly.js
/**
 * Singly Linked List Node
 * @typedef {Object} Node
 * @property {number} value
 * @property {Node} next
 */
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }
}

/**
 * Singly Linked List
 * @typedef {Object} LinkedList
 * @property {Node} head
 * @property {number} size
 */
class LinkedList {
    constructor(value = null) {
        this.head = (value == null) ? null : new Node(value);
        this.size = (this.head == null) ? 0 : 1;
    }


    /**
     * Inserts a new node as the head of the linked list  
     * `O(1) time`
     * @param {number} value 
     * @returns {boolean}
     */
    insertHead(value) {
        if (value == null || value == undefined) {
            console.warn('Missing value during head insertion');
            return false;
        }

        try {
            const newHead = new Node(value);
            newHead.next = this.head;
            this.head = newHead;
            this.size += 1;
            return true;
        } catch (error) {
            console.error(`Error occurred while inserting a new head node: ${error.message}`)
            throw error;
        }
    }

    
    /**
     * Inserts a new node as the tail of the linked list  
     * `O(n) time`
     * @param {number} value 
     * @returns {boolean}
     */
    insertTail(value) {
        if (value == null || value == undefined) {
            console.warn('Missing value during tail insertion');
            return false;
        }

        try {
            if (this.size == 0) {
                return this.insertHead(value);
            }
            
            let curr = this.head;

            while (curr != null && curr.next != null) {
                curr = curr.next;
            }

            const newTail = new Node(value);
            curr.next = newTail;
            this.size += 1;
            return true;
        } catch (error) {
            console.error(`Error occurred while inserting a new tail node: ${error.message}`)
            throw error;
        }
    }
    
    
    /**
     * Inserts a new node into the given index
     * `O(n) time`
     * @param {number} value 
     * @param {number} index 
     * @returns {boolean}
     */
    insertAtIndex(value, index) {
        if (value == null || value == undefined) {
            console.warn('Missing value during index insertion');
            return false;
        }
        
        if (index == null || index == undefined) {
            console.warn('Missing index during index insertion');
            return false;
        }

        if (index < 0 || index >= this.size) {
            console.warn('Invalid index during index insertion');
            return false;
        }

        try {
            if (index == 0) {
                return this.insertHead(value);
            }

            if (index == this.size - 1) {
                return this.insertTail(value);
            }

            let curr = this.head;

            for (let i = 0; i < index - 1; i++) {
                curr = curr.next;
            }

            const newIdxNode = new Node(value);
            const prevIdxNode = curr.next;

            curr.next = newIdxNode;
            newIdxNode.next = prevIdxNode;
            this.size += 1;
            return true;
        } catch (error) {
            console.error(`Error occurred while inserting a new node at index ${index}: ${error.message}`)
            throw error;
        }
    }

    
    /**
     * Deletes the head node
     * `O(1) time`
     * @returns {boolean}
     */
    deleteHead() {
        if (this.head == null) {
            return false;
        }

        try {
            const newHead = this.head.next;
            this.head.next = null;
            this.head = newHead;
            this.size -= 1;
            return true;
        } catch (error) {
            console.error(`Error occurred while deleting head node: ${error.message}`)
            throw error;
        }
    }
    

    /**
     * Deletes the tail node
     * `O(n) time`
     * @returns {boolean}
     */
    deleteTail() {
        if (this.head == null) {
            return false;
        }

        if (this.size == 1) {
            this.head = null;
            this.size -= 1;
            return true;
        }

        try {
            let curr = this.head;

            for (let i = 0; i < this.size - 2; i++) {
                curr = curr.next;
            }

            curr.next = null;
            this.size -= 1;
            return true;
        } catch (error) {
            console.error(`Error occurred while deleting head node: ${error.message}`)
            throw error;
        }
    }


    /**
     * Deletes a node at the given index
     * `O(n) time`
     * @param {number} index 
     * @returns {boolean}
     */
    deleteAtIndex(index) {
        if (index == null || index == undefined) {
            console.warn('Missing index during index deletion');
            return false;
        }

        if (index < 0 || index > this.size - 1) {
            console.warn('Invalid index during index deletion');
            return false;
        }

        try {
            if (index == 0) {
                return this.deleteHead();
            }

            if (index == this.size - 1) {
                return this.deleteTail();
            }

            let curr = this.head;

            for (let i = 0; i < index - 1; i++) {
                curr = curr.next;
            }

            const deletedNode = curr.next;
            const nextNode = deletedNode.next;

            deletedNode.next = null;
            curr.next = nextNode;
            this.size -= 1;
            return true;
        } catch (error) {
            console.error(`Error occurred while deleting a node at index ${index}: ${error.message}`)
            throw error;
        }
    }
   

    /**
     * Returns if the list contains the given value
     * `O(n) time`
     * @param {number} value - value to search for
     * @returns {boolean}
     */
    hasValue(value) {
        if (value == null || value == undefined) {
            console.warn('Invalid search value');
            return false;
        }

        if (this.size == 0) {
            return false;
        }

        try {
            let curr = this.head;

            while (curr != null) {
                console.log("Current node value:", curr.value); // Print current value
                if (curr.value == value) {
                    console.log("Value found!"); // Print if value is found
                    return true;
                }
                curr = curr.next;
            }
            console.log("Value not found."); // Print if value is not found
            return false;
        } catch (error) {
            console.error(`Error occurred while searching for value: ${error.message}`)
            throw error;
        }
    }
    

    /**
     * Returns the size of the linked list
     * `O(1) time`
     * @returns {number}
     */
    getSize() {
        return this.size;
    }


    /**
     * Prints the linked list to the console
     * `O(n) time`
     */
    printList() {
        if (this.head === null) {
            console.log("Empty List");
            return;
        }

        let curr = this.head;
        let output = '\nLinked List: ';
        while (curr != null) {
            output += `${curr.value} -> `;
            curr = curr.next;
        }
        output += 'NULL\n';
        console.log(output);
    }
}


module.exports = { LinkedList }

